### PR TITLE
ConfigMap Volume

### PR DESCRIPTION
**Pull request by pmorie**
Do you like volumes?  Do you like ConfigMaps?  If so, this is the PR for you.

### CODE REVIEW

**Comment by bgrant0607** [comment]
Is the move to api/v1 only in this PR, or is there another PR for that?

**Comment by pmorie** [comment]
@bgrant0607 API move was in #19716, this was implemented on top of that one.

**Comment by pmorie** [comment]
@thockin to limit the size of this PR, I'm going to open other for extracting the atomic write object, then rebase this onto that one to use it.  Alternately, we could review and merge this one as is and I will add the updating feature in a subsequent PR

**Comment by mwielgus** [comment]
We need this PR for custom metrics support in 1.2. What is the current status?

cc: @piosz @fgrzadkowski @erictune @davidopp

**Comment by pmorie** [comment]
@mwielgus Apologies, I missed the ping on this -- it's been blocked on extracting the atomic update logic for re-use outside the downward API volume.  Should be ready for final review now.

**Comment by pmorie** [comment]
@thockin PTAL, ready for final review

File: pkg/api/types.go [hunk 1]
```diff
@@ -688,6 +690,36 @@ type AzureFileVolumeSource struct {
...
+// The contents of the target ConfigMap's Data field will be presented in a
+// volume as files using the keys in the Data field as the file names, unless
+// the items element is populate with specific mappings of keys to paths.
+// ConfigMap volumes support ownership management and SELinux relabeling.
+type ConfigMapVolumeSource struct {
+	LocalObjectReference `json:",inline"`
+	// If unspecified, each key-value pair in the Data field of the
+	// referenced ConfigMap will be projected into the volume as a file whose name
+	// is the key and content is the value.
+	// If specified, the listed keys will be project into the specified paths, and
```
Comment (ncdc) [inline pkg/api/types.go]: will be projected

File: pkg/api/types.go [hunk 2]
```diff
@@ -688,6 +690,36 @@ type AzureFileVolumeSource struct {
...
+	// unlisted keys will not be present.
+	// If a key is specified with is not present in the ConfigMap, the volume setup
```
Comment (ncdc) [inline pkg/api/types.go]: which is not present

File: pkg/volume/configmap/configmap.go [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+	kubeClient := b.plugin.host.GetKubeClient()
+	if kubeClient == nil {
+		return fmt.Errorf("Cannot setup configMap volume %v because kube client is not configured", b.volName)
+	}
+
+	configMap, err := kubeClient.Core().ConfigMaps(b.pod.Namespace).Get(b.source.Name)
+	if err != nil {
+		glog.Errorf("Couldn't get configMap %v/%v", b.pod.Namespace, b.source.Name)
+		return err
+	} else {
```
Comment (ncdc) [inline pkg/volume/configmap/configmap.go]: `else` not needed?

File: pkg/volume/configmap/configmap.go [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+	}
+
+	kubeClient := b.plugin.host.GetKubeClient()
+	if kubeClient == nil {
+		return fmt.Errorf("Cannot setup configMap volume %v because kube client is not configured", b.volName)
+	}
+
+	configMap, err := kubeClient.Core().ConfigMaps(b.pod.Namespace).Get(b.source.Name)
+	if err != nil {
+		glog.Errorf("Couldn't get configMap %v/%v", b.pod.Namespace, b.source.Name)
```
Comment (ncdc) [inline pkg/volume/configmap/configmap.go]: add %v for err to the log message?

File: pkg/api/v1/types.go [hunk 5]
```diff
@@ -808,6 +810,35 @@ type AzureFileVolumeSource struct {
...
+}
+
+// Maps a string key to a path within a volume.
+type KeyToPath struct {
+	// The key to project.
+	Key string `json:"key"`
+
+	// The relative path of the file to map the key to.
+	// May not begin with '/'.
+	// May not contain the string '..'.
```
Comment (thockin) [inline pkg/api/v1/types.go]: This comment is imprecise.

File: pkg/api/types.go [hunk 6]
```diff
@@ -688,6 +690,35 @@ type AzureFileVolumeSource struct {
...
+}
+
+// Maps a string key to a path within a volume.
+type KeyToPath struct {
+	// The key to project.
+	Key string `json:"key"`
+
+	// The relative path of the file to map the key to.
+	// May not begin with '/'.
+	// May not contain the string '..'.
```
Comment (thockin) [inline pkg/api/types.go]: This comment is imprecise.

**Comment by thockin** [comment]
2 small comment clarifications and LGTM.  I am OOO this week, so I am assigning to Minhan to final-approve.

File: pkg/api/types.go (hunk 6 above)
Comment (thockin-cc) [inline pkg/api/types.go]: Also, squash please

**Comment by pmorie** [comment]
@thockin @freehan rebased, but I decline to squash because the second commit is generated stuff which is a pain to rebase, not to mention more layerlike than many of the commits in multi-commit merges in the last hundred commits or so...

**Comment by thockin-cc** [comment]
The problem with not squashing is that there is a point in the commit
history that won't compile, which makes things like git bisect sort of
awful.  I have a few minutes now, will take a look.

**Comment by pmorie** [comment]
@thockin-cc I thought the guidance on squashing specifically says that it's not necessary to ensure that the build works after each commit -- I do see your point, but have never squashed regeneration before for the reasons I mentioned earlier.  If it is really critical to maintaining that invariant, we should invest in having some tooling that makes it easier to rebase all this generated code.

**Comment by pmorie** [comment]
@thockin what about making the merge-bot able to squash certain types of commits before final merge? Hrm...

**Comment by thockin** [comment]
It's fine.  We are not exactly consistent on it.

**Comment by pmorie** [comment]
@thockin I think I can get away with it now, if you have a burning desire.  LMK.

Your humble servant, 

P

**Comment by thockin-cc** [comment]
Don't touch, let the mergebot do its thing

**Comment by pmorie** [comment]
:point_right: :point_right:

**Comment by erictune** [comment]
The real problem is not so much squashing or not.  It is checking in
generated files.

**Pull request merged**